File: python/leetcode/284. Peeking Iterator.py
# Below is the interface for Iterator, which is already defined for you.
#
# class Iterator:
#     def __init__(self, nums):
#         """
#         Initializes an iterator object to the beginning of a list.
#         :type nums: List[int]
#         """
#
#     def hasNext(self):
#         """
#         Returns true if the iteration has more elements.
#         :rtype: bool
#         """
#
#     def next(self):
#         """
#         Returns the next element in the iteration.
#         :rtype: int
#         """


class PeekingIterator:
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self.pos = -1
        self.nums = []
        self.iterator = iterator

        while iterator.hasNext():
            self.nums.append(iterator.next())

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        if self.hasNext():
            return self.nums[self.pos + 1]

    def next(self):
        """
        :rtype: int
        """
        if self.hasNext():
            self.pos += 1
            return self.nums[self.pos]

    def hasNext(self):
        """
        :rtype: bool
        """
        if len(self.nums) < 1:
            return False

        return self.pos + 1 < len(self.nums)


# Your PeekingIterator object will be instantiated and called as such:
# iter = PeekingIterator(Iterator(nums))
# while iter.hasNext():
#     val = iter.peek()   # Get the next element but not advance the iterator.
#     iter.next()         # Should return the same value as [val].

File: c++/leetcode/284. Peeking Iterator.cpp
// Below is the interface for Iterator, which is already defined for you.
// **DO NOT** modify the interface for Iterator.

class Iterator
{
  struct Data;
  Data *data;

public:
  Iterator(const vector<int> &nums);
  Iterator(const Iterator &iter);
  virtual ~Iterator();
  // Returns the next element in the iteration.
  int next();
  // Returns true if the iteration has more elements.
  bool hasNext() const;
};

class PeekingIterator : public Iterator
{
  int pos;
  vector<int> v;

public:
  PeekingIterator(const vector<int> &nums) : Iterator(nums), v{nums}
  {
    // Initialize any member here.
    // **DO NOT** save a copy of nums and manipulate it directly.
    // You should only use the Iterator interface methods.
    pos = -1;
  }

  // Returns the next element in the iteration without advancing the iterator.
  int peek()
  {
    if (hasNext())
      return v[pos + 1];
    return -1;
  }

  // hasNext() and next() should behave the same as in the Iterator interface.
  // Override them if needed.
  int next()
  {
    if (hasNext())
      return v[++pos];
    return -1;
  }

  bool hasNext() const
  {
    if (v.size() < 1)
      return false;
    return pos + 1 < v.size();
  }
};
